File: src/simulation/deck.rs
use crate::card::{Card, CardDatabase, CardDatabaseError};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum DeckError {
    #[error("IO error: {0}")]
    IoError(#[from] std::io::Error),
    #[error("Invalid deck format at line {line}: {reason}")]
    InvalidFormat { line: usize, reason: String },
    #[error("Card database error: {0}")]
    DatabaseError(#[from] CardDatabaseError),
}

/// Parse a deck file and return expanded list of cards
/// Format: "4 Card Name" per line, supports comments with # or //
pub fn parse_deck_file(
    path: &str,
    database: &CardDatabase,
) -> Result<Vec<Card>, DeckError> {
    let content = std::fs::read_to_string(path)?;
    let mut deck = Vec::new();

    for (line_num, line) in content.lines().enumerate() {
        let trimmed = line.trim();

        // Skip empty lines and comments
        if trimmed.is_empty() || trimmed.starts_with('#') || trimmed.starts_with("//") {
            continue;
        }

        // Parse "N Card Name" format
        let parts: Vec<&str> = trimmed.splitn(2, ' ').collect();
        if parts.len() != 2 {
            return Err(DeckError::InvalidFormat {
                line: line_num + 1,
                reason: "Expected format: 'COUNT CARD_NAME'".to_string(),
            });
        }

        let count_str = parts[0];
        let card_name = parts[1].trim();

        let count: usize = count_str.parse().map_err(|_| DeckError::InvalidFormat {
            line: line_num + 1,
            reason: format!("'{}' is not a valid number", count_str),
        })?;

        // Get card from database
        let card = database.get_card(card_name)?;

        // Add card 'count' times
        for _ in 0..count {
            deck.push(card.clone());
        }
    }

    Ok(deck)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_deck_file() {
        let db = CardDatabase::from_file("cards.json").expect("Failed to load cards");
        let deck = parse_deck_file("deck.txt", &db).expect("Failed to parse deck");
        
        // deck.txt should have 60 cards total
        assert_eq!(deck.len(), 60, "Deck should have 60 cards");
    }

    #[test]
    fn test_deck_expansion() {
        let db = CardDatabase::from_file("cards.json").expect("Failed to load cards");
        let deck = parse_deck_file("deck.txt", &db).expect("Failed to parse deck");
        
        // Count Forest cards (should be 2)
        let forest_count = deck.iter().filter(|c| c.name() == "Forest").count();
        assert_eq!(forest_count, 2, "Should have 2 Forest cards");
    }

    #[test]
    fn test_invalid_card_name() {
        let db = CardDatabase::from_file("cards.json").expect("Failed to load cards");
        let result = parse_deck_file("deck.txt", &db);
        
        // This should succeed since deck.txt has valid cards
        assert!(result.is_ok());
    }
}

File: java/src/main/java/com/mtg/reanimator/simulation/Deck.java
package com.mtg.reanimator.simulation;

import com.mtg.reanimator.card.Card;
import com.mtg.reanimator.card.CardDatabase;
import com.mtg.reanimator.card.CardDatabaseException;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Deck representation and parsing.
 * Matches the Rust deck.rs implementation.
 */
public class Deck {
    private final List<Card> cards;
    private final String name;

    public Deck(List<Card> cards, String name) {
        this.cards = new ArrayList<>(cards);
        this.name = name;
    }

    /**
     * Load a deck from a file.
     * Format: "4 Card Name" per line, supports comments with # or //
     *
     * @param path Path to the deck file
     * @param db   Card database
     * @return Parsed deck
     * @throws DeckException if parsing fails
     */
    public static Deck loadFromFile(String path, CardDatabase db) throws DeckException {
        try {
            String content = Files.readString(Path.of(path));
            List<Card> cards = new ArrayList<>();
            String[] lines = content.split("\n");

            for (int lineNum = 0; lineNum < lines.length; lineNum++) {
                String line = lines[lineNum].trim();

                // Skip empty lines and comments
                if (line.isEmpty() || line.startsWith("#") || line.startsWith("//")) {
                    continue;
                }

                // Parse "N Card Name" format
                int spaceIdx = line.indexOf(' ');
                if (spaceIdx == -1) {
                    throw new DeckException("Invalid deck format at line " + (lineNum + 1)
                            + ": Expected format 'COUNT CARD_NAME'");
                }

                String countStr = line.substring(0, spaceIdx);
                String cardName = line.substring(spaceIdx + 1).trim();

                int count;
                try {
                    count = Integer.parseInt(countStr);
                } catch (NumberFormatException e) {
                    throw new DeckException("Invalid deck format at line " + (lineNum + 1)
                            + ": '" + countStr + "' is not a valid number");
                }

                // Get card from database
                try {
                    Card card = db.getCard(cardName);
                    for (int i = 0; i < count; i++) {
                        cards.add(card);
                    }
                } catch (CardDatabaseException e) {
                    throw new DeckException("Card not found at line " + (lineNum + 1) + ": " + cardName);
                }
            }

            // Extract deck name from path
            String fileName = Path.of(path).getFileName().toString();
            String deckName = fileName.endsWith(".txt")
                    ? fileName.substring(0, fileName.length() - 4)
                    : fileName;

            return new Deck(cards, deckName);
        } catch (IOException e) {
            throw new DeckException("Failed to read deck file: " + e.getMessage());
        }
    }

    public List<Card> getCards() {
        return new ArrayList<>(cards);
    }

    public int size() {
        return cards.size();
    }

    public String getName() {
        return name;
    }

    /**
     * Exception thrown when deck parsing fails.
     */
    public static class DeckException extends Exception {
        public DeckException(String message) {
            super(message);
        }
    }
}

